### PR TITLE
Guidance on Posit Workbench and Kubernetes

### PR DESCRIPTION
**Pull request by terrymclaughlin**
This document aims to provide users of Posit Workbench with high-level information on

- what Kubernetes is,
- why Posit Workbench has been deployed using Kubernetes and how it works,
- the importance of optimal use of cloud computing resources in respect of financial cost to PHS, and
- best practice for using Posit Workbench deployed using Kubernetes.

I would welcome any feedback and suggested changes on this document.  My intention is for this document to make available through the Data Science Knowledge Base.

### CODE REVIEW

File: posit_workbench_and_kubernetes.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+The reasons that Posit Workbench has been deployed using Kubernetes, and AKS specifically, are to
+ 
+- isolate sessions from one another,
+- enable autoscaling, and thus
+- facilitate optimal usage of cloud computing resources.
+
+We need to introduce some Kubernetes concepts at this point to explain how Kubernetes delivers this.
+
+### Pods, Nodes and Clusters, oh my!
```
[the remark this answers is not in the document]
Comment (rmccreath) [reply posit_workbench_and_kubernetes.md:48]: 😍 this is the kind of guidance writing we need!

**Review by rmccreath — COMMENTED** [review]
This is brilliant, @terrymclaughlin - I think the level and tone are set really well, I especially like the breakdown of memory recommendations. I've made some suggestions and raised a couple of questions, absolutely nothing of any urgency and would be more than happy for this to be live as is.

File: posit_workbench_and_kubernetes.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrymclaughlin) [reply posit_workbench_and_kubernetes.md:48]: 😁

File: posit_workbench_and_kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+When a node no longer contains any running pods, after a defined period of time, the node shuts down and is removed from the cluster.  A minimum of one node will always be left running.
+
+From a user experience perspective, it will take longer for a Posit Workbench session to start and be ready to use when a new node needs to be started up.  This can take several minutes.  However, in most circumstances, a Posit Workbench session will start on an already running node, and this will be ready for use in just a few seconds.
+
+### Isolated sessions
+
+Containerising (yes, that's a word!) Posit Workbench has the added advantage of isolating users' sessions from one another.  A session is only permitted to consume CPU time and memory up to the limits set by Kubernetes, or requested by the user (if they are less than those imposed by Kubernetes).  Therefore, a user running a CPU and/or memory intensive R script will not result in a poor experience for other users.
+
+### Optimal usage of cloud computing resources
```
[the remark this answers is not in the document]
Comment (terrymclaughlin) [reply posit_workbench_and_kubernetes.md]: Changed to h2

File: posit_workbench_and_kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- Highlighted in green are the CPU and memory parameters for your session. These can be configured to suit the compute demands of your script.
+- Highlighted in blue are the maximum values for CPU and memory that you are permitted to request for your session.  These values are dependent on the profile that you have been assigned to (described in more detail below).
+
+#### CPUs ####
+
+The Central Processing Unit (CPU) is the primary component of a computer that executes instructions.
+
+Modern CPUs can have multiple *cores*, and on those cores, multiple *threads* can be executed simultaneously.
+
+On cloud computing platforms, we refer to virtual CPUs (vCPUs) where 1 vCPU equates to 1 thread of execution on a physical core.  Kubernetes adds a layer of abstraction such that a pod can execute in multiple threads.
```
[the remark this answers is not in the document]
Comment (terrymclaughlin) [reply posit_workbench_and_kubernetes.md:94]: @rmccreath 

The Kubernetes documentation on "[Resource Management for Pods and Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/)" describes how resources are allocated to a pod.  Whilst it doesn't explicitly talk about how this works in a cloud environment, it's a probably a good resource for people wanting to know more.

File: posit_workbench_and_kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+There are ways to reduce memory consumption, by performing aggregation within the database, performing operations on chunks of data, or using technologies such as Apache Arrow, for working with larger-than-memory datasets stored on disk. *Guidance to be written on this*
+
+Again, the less memory you request, the greater number of pods running Posit Workbench sessions can be squeezed onto a single node, thus minimising cost to PHS.
+
+#### Profiles ####
+
+Your user account is, by default, assigned to a generic profile which grants you permission to request up to x CPU and x MB of memory for your session.
+
+If the work that you do requires a greater number of CPUs or amount of memory than the defaults, you can request that your user account is assigned to an enhanced profile.  This request should be submitted in writing to ... explaining the reasons you require greater resources and for how long. *TBC*
```
[the remark this answers is not in the document]
Comment (terrymclaughlin) [reply posit_workbench_and_kubernetes.md:139]: Agreed

[the remark this answers is not in the document]
Comment (terrymclaughlin) [reply posit_workbench_and_kubernetes.md]: [Write guidance on requesting a different Kubernetes profile #29](https://github.com/Public-Health-Scotland/R-Resources/issues/29)

**Review by rmccreath — APPROVED** [review]

**Pull request merged**